community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/RelationshipTypeCheckerStep.java

```diff
@@ -0,0 +1,142 @@
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.unsafe.impl.batchimport;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.neo4j.kernel.impl.store.record.RelationshipRecord;
+import org.neo4j.unsafe.impl.batchimport.input.InputRelationship;
+import org.neo4j.unsafe.impl.batchimport.staging.BatchSender;
+import org.neo4j.unsafe.impl.batchimport.staging.Configuration;
+import org.neo4j.unsafe.impl.batchimport.staging.ProcessorStep;
+import org.neo4j.unsafe.impl.batchimport.staging.StageControl;
+import org.neo4j.unsafe.impl.batchimport.store.BatchingTokenRepository.BatchingRelationshipTypeTokenRepository;
+
+/**
+ * Counts relationships per type to later be able to provide all types, even sorted in descending order
+ * of number of relationships per type.
+ */
+public class RelationshipTypeCheckerStep extends ProcessorStep<Batch<InputRelationship,RelationshipRecord>>
+{
+    private static final Comparator<Map.Entry<Object,AtomicLong>> SORT_BY_COUNT_DESC =
+            (e1,e2) -> Long.compare( e2.getValue().get(), e1.getValue().get() );
+    private static final Comparator<Map.Entry<Object,AtomicLong>> SORT_BY_ID_DESC =
+            (e1,e2) -> Integer.compare( (Integer)e2.getKey(), (Integer)e1.getKey() );
+    private final Map<Object,AtomicLong> allTypes = new HashMap<>();
+    private final BatchingRelationshipTypeTokenRepository typeTokenRepository;
+    private Map.Entry<Object,AtomicLong>[] sortedTypes;
+    private long totalCount;
+
+    public RelationshipTypeCheckerStep( StageControl control, Configuration config,
+            BatchingRelationshipTypeTokenRepository typeTokenRepository )
+    {
+        super( control, "TYPE", config, 0 );
+        this.typeTokenRepository = typeTokenRepository;
+    }
+
+    @Override
+    protected void process( Batch<InputRelationship,RelationshipRecord> batch, BatchSender sender ) throws Throwable
+    {
+        for ( InputRelationship relationship : batch.input )
+        {
+            Object type = relationship.typeAsObject();
+            AtomicLong count = allTypes.get( type );
+            // Check w/o synchronized, it's fine
```

Comment:
If the `get` effectively performed a volatile read, and the `put` effectively performed a volatile write, then there would be no problem with this kind of double-checked locking.